### PR TITLE
refactor (TCOMP-294) : docker image rest api

### PR DESCRIPTION
**Pull request by sgandon**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
The rest api was located in a module called **web-test** and was using json-io for converting all objects
the service module was named **component-api-service** and was mixing OSGI and Spring services along with all their dependencies.

**What is the new behavior?**
the rest/spring api is now located in a module named **services/components-api-service-rest**
the OSGI api is located in a module named **services/components-api-service-osgi**
There is a common module to both previous modules called **services/components-api-service-common**

**Does this PR introduce a breaking change?**

- [X] Yes
- [ ] No

* moved test utilities classes from **components-api-service** to **components-api**, so pom need to be updated
* moved the spring test application *SpringTestApp* from **components-api-service** to **services/components-api-service-rest**, so pom needs to be updated.

**Other information**:
For adding a spring integration test into a component, you need to add these depednencies
```
       <dependency>
            <groupId>org.talend.components</groupId>
            <artifactId>components-api-service-rest</artifactId>
            <version>0.15.0-SNAPSHOT</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.talend.components</groupId>
            <artifactId>components-api-service-rest</artifactId>
            <version>0.15.0-SNAPSHOT</version>
            <type>test-jar</type>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-test</artifactId>
            <scope>test</scope>
        </dependency>

```

### CODE REVIEW

**Comment by sgandon** [comment]
I still have to do some clean up and make the *service-common* classes embedded in the OSGI bundle for the service implementation.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/224/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/224/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/226/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/226/)

File: components-api/pom.xml [hunk 1]
```diff
@@ -61,9 +61,9 @@
             <scope>test</scope>
         </dependency>
         <dependency>
-			<groupId>org.talend.components</groupId>
-			<artifactId>components-api-full-example</artifactId>
-			<version>0.1.0</version>     
+            <groupId>org.talend.components</groupId>
+            <artifactId>components-api-full-example</artifactId>
```
Comment (talend-glorieux) [inline components-api/pom.xml]: Extra spaces at EOL

File: components-jira/pom.xml [hunk 2]
```diff
@@ -51,40 +51,38 @@
...
-                    <artifactId>logback-classic</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>org.slf4j</groupId>
-                    <artifactId>log4j-over-slf4j</artifactId>
-                </exclusion>
-            </exclusions>
         </dependency>
-
+
```
Comment (talend-glorieux) [inline components-jira/pom.xml]: Extra spaces

File: services/components-api-service-common/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/target/
+/target/
```
Comment (talend-glorieux) [inline services/components-api-service-common/.gitignore]: Why is this added twice?

File: services/components-api-service-common/readme.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+```
+mvn spring-boot:run
+```
+
+run docker: 
+
+```
+mvn -Pdocker install
+docker -p 8080:8080 run talend/components-api-service-rest:0.15.0-SNAPSHOT
```
Comment (talend-glorieux) [inline services/components-api-service-common/readme.md]: I guess we should have some Docker images naming conventions.

File: services/components-api-service-rest/src/main/java/org/talend/components/service/rest/JsonSchema2HttpMessageConverter.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            try {
+                return JsonUtil.fromJson(inputMessage.getBody());
+            } catch (Exception e) {
+                throw new IOException(e.getMessage());
+            }
+        } else {
+            return JsonReader.jsonToJava(inputMessage.getBody(), null);
+        }
+    }
+}
```
Comment (talend-glorieux) [inline services/components-api-service-rest/src/main/java/org/talend/components/service/rest/JsonSchema2HttpMessageConverter.java]: No newline at EOF.

File: services/pom.xml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+				</configuration>
+			</plugin>
+			<plugin>
+				<groupId>org.codehaus.mojo</groupId>
+				<artifactId>versions-maven-plugin</artifactId>
+				<version>2.2</version>
+			</plugin>
+		</plugins>
+	</build>
+</project>
```
Comment (talend-glorieux) [inline services/pom.xml]: No newline at EOF.

**Comment by talend-glorieux** [comment]
Where can I find the Dockerfile used for the build?

**Comment by sgandon** [comment]
the docker file is generated from the pom.xml on the fly. You'll find it here *services/components-api-service-rest-all-components/target/docker/Dockerfile* if you build the artifact yourself.
here is what it looks like 
```
FROM mykro/java8-jre:latest
ADD /components-api-service-rest-all-components-0.1.0-SNAPSHOT.jar //
CMD java -Dfile.encoding=UTF-8 -jar /components-api-service-rest-all-components-0.1.0-SNAPSHOT.jar
```

**Comment by talend-glorieux** [comment]
Looks like a single guy is maintaining mykro/java8-jre? It's not really safe.
I can't seem to find an official Oracle Java docker images but there's already a [java image on docker-talend](https://github.com/Talend/docker-talend/tree/master/java). Could this be a suitable replacement?

**Comment by sgandon** [comment]
I must admit I did not take a look at that and used Ryans proposal to have something working for you @talend-glorieux  quickly ;) 
I'll probably look at what dataprep uses to be homogeneous.

File: services/components-api-service-rest/src/main/java/org/talend/components/service/rest/ComponentServiceRest.java [hunk 7]
```diff
@@ -10,7 +10,7 @@
 // 9 rue Pages 92150 Suresnes, France
 //
 // ============================================================================
-package org.talend.components.api.service.internal.spring;
+package org.talend.components.service.rest;
```
Comment (vjaouen) [inline services/components-api-service-rest/src/main/java/org/talend/components/service/rest/ComponentServiceRest.java]: typo lines 345 & 346 : Connetor --> Connector

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/239/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/239/)

**Comment by sgandon** [comment]
@talend-glorieux, in fact *mykro/java8-jre* is the base docker image also used by dataprep. I don't know if there is Talend standard, we should probably ask to the Devops.

File: services/components-api-service-common/readme.md (hunk 4 above)
Comment (sgandon) [inline services/components-api-service-common/readme.md]: In fact we have some.
https://in.talend.com/12813184
I'll update the PR accordingly.

**Comment by sgandon** [comment]
I'll merge right now and fix Geoffroy issue later, cause I'd like to have this on master soon.
